contracts/ics721/src/ibc.rs

```diff
@@ -0,0 +1,574 @@
+use cw20_ics20::state::ChannelInfo;
+use cw721::Cw721ExecuteMsg;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{
+    attr, entry_point, from_binary, to_binary, Binary, ContractResult, DepsMut, Empty, Env,
+    IbcBasicResponse, IbcChannel, IbcChannelCloseMsg, IbcChannelConnectMsg, IbcChannelOpenMsg,
+    IbcEndpoint, IbcOrder, IbcPacket, IbcPacketAckMsg, IbcPacketReceiveMsg, IbcPacketTimeoutMsg,
+    IbcReceiveResponse, Reply, Response, SubMsg, WasmMsg,
+};
+
+use crate::error::{ContractError, Never};
+use crate::state::{CHANNEL_INFO, CHANNEL_STATE};
+
+pub const ICS721_VERSION: &str = "ics721-1";
+pub const ICS721_ORDERING: IbcOrder = IbcOrder::Unordered;
+
+// TODO: need to define proto for chain to parse this?
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug, Default)]
+pub struct Ics721Packet {
+    /// uniquely identifies the collection to which the NFT belongs
+    /// the sg721 collection contract address
+    pub class_id: String,
+    // a link for class/contract-level metadata
+    /// https://docs.opensea.io/docs/contract-level-metadata
+    pub class_uri: Option<String>,
+    /// uniquely identifies NFTs within the collection that is being transferred
+    pub token_ids: Vec<String>,
+    /// https://docs.opensea.io/docs/metadata-standards
+    pub token_uris: Vec<String>,
+    pub sender: String,
+    pub receiver: String,
+}
+
+impl Ics721Packet {
+    pub fn new(
+        class_id: &str,
+        class_uri: Option<&str>,
+        token_ids: Vec<&str>,
+        token_uris: Vec<&str>,
+        sender: &str,
+        receiver: &str,
+    ) -> Self {
+        Ics721Packet {
+            class_id: class_id.to_string(),
+            class_uri: class_uri.map(str::to_string),
+            token_ids: token_ids
+                .iter()
+                .map(|&s| s.to_string())
+                .collect::<Vec<String>>(),
+            token_uris: token_uris
+                .iter()
+                .map(|&s| s.to_string())
+                .collect::<Vec<String>>(),
+            sender: sender.to_string(),
+            receiver: receiver.to_string(),
+        }
+    }
+
+    pub fn validate(&self) -> Result<(), ContractError> {
+        // TODO:
+        // https://github.com/public-awesome/contracts/issues/53
+        // validate class_id is a contract address
+        // validate class_uri is a uri if it exists
+        // validate same number of token_ids and token_uris
+        Ok(())
+    }
+}
+
+// TODO: copy pasta, make parent public
+// create a serialized success message
+fn ack_success() -> Binary {
+    let res = Ics20Ack::Result(b"1".into());
+    to_binary(&res).unwrap()
+}
+
+// TODO: copy pasta, make parent public
+// create a serialized error message
+fn ack_fail(err: String) -> Binary {
+    let res = Ics20Ack::Error(err);
+    to_binary(&res).unwrap()
+}
+
+const SEND_NFT_ID: u64 = 1338;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
```

Comment:
reply logic confused me for some time like, *is there an reply method for ibc?* 
Better to move this into contract.rs.